code/controllers/subsystem/overlays.dm

```diff
@@ -95,167 +38,83 @@ SUBSYSTEM_DEF(overlays)
 	iconbro.icon = icon
 	return iconbro.appearance
 
-// If the overlay has a planeset (e.g., emissive), mark for ZM mangle. This won't catch overlays on overlays, but the flag can just manually be set in that case.
-#define ZM_AUTOMANGLE(target) if ((target):plane != FLOAT_PLANE) { SSoverlays.context_needs_automangle = TRUE; }
-
-/atom/proc/build_appearance_list(old_overlays)
-	var/static/image/appearance_bro = new()
-	var/list/new_overlays = list()
-
-	for (var/overlay in (islist(old_overlays) ? old_overlays : list(old_overlays)))
+/atom/proc/build_appearance_list(build_overlays)
+	if (!islist(build_overlays))
+		build_overlays = list(build_overlays)
+	for (var/overlay in build_overlays)
 		if(!overlay)
+			build_overlays -= overlay
 			continue
 		if (istext(overlay))
-			new_overlays += iconstate2appearance(icon, overlay)
+#ifdef UNIT_TESTS
+			// This is too expensive to run normally but running it during CI is a good test
+			if(!skip_sprite_error)
+				var/list/icon_states_available = icon_states(icon)
+				if(!(overlay in icon_states_available))
+					var/icon_file = "[icon]" || "Unknown Generated Icon"
+					stack_trace("Invalid overlay: Icon object '[icon_file]' [REF(icon)] used in '[src]' [type] is missing icon state [overlay].")
+					continue
+#endif
+			build_overlays -= overlay
+			build_overlays += iconstate2appearance(icon, overlay)
 		else if(isicon(overlay))
-			new_overlays += icon2appearance(overlay)
-		else
-			if(isloc(overlay))
-				var/atom/A = overlay
-				if (A.flags_1 & OVERLAY_QUEUED_1)
-					COMPILE_OVERLAYS(A)
-			appearance_bro.appearance = overlay //this works for images and atoms too!
-			if(!ispath(overlay))
-				var/image/I = overlay
-				appearance_bro.dir = I.dir
-			new_overlays += appearance_bro.appearance
-	return new_overlays
+			build_overlays -= overlay
+			build_overlays += icon2appearance(overlay)
+	return build_overlays
 
-// The same as the above, but with ZM_AUTOMANGLE.
-/atom/movable/build_appearance_list(old_overlays)
-	var/static/image/appearance_bro = new()
-	var/list/new_overlays = list()
+#ifdef UNIT_TESTS
+/atom/var/skip_sprite_error
+/obj/item/gun/skip_sprite_error = TRUE
+// I hate this but gun sprites are not implemented properly, and unit test blames this
+// We'll need to clean up gun sprites someday, but not right now, it's a mess
+#endif
```

Comment:
Error details can be found in:
* https://github.com/BeeStation/BeeStation-Hornet/actions/runs/8277897976/job/22649152337